core/cache/modelsummarywatcher_test.go

```diff
@@ -0,0 +1,578 @@
+			Name:             "model-1",
+			Admins:           []string{"test-admin"},
+			Status:           cache.StatusGreen,
+			MachineCount:     1,
+			ApplicationCount: 1,
+			UnitCount:        1,
+		}, {
+			UUID:      "model-2-uuid",
+			Namespace: "bob",
+			Name:      "model-2",
+			Admins:    []string{"bob"},
+			Status:    cache.StatusGreen,
+		},
+	})
+}
+
+func (s *modelSummaryWatcherSuite) TestInitialModelsCharlie(c *gc.C) {
+	watcher := s.controller.WatchModelsAsUser("charlie")
+	defer workertest.CleanKill(c, watcher)
+
+	changes := watcher.Changes()
+	initial := s.next(c, changes)
+	c.Assert(initial, gc.HasLen, 0)
+}
+
+func (s *modelSummaryWatcherSuite) TestAddPermissionShowsModel(c *gc.C) {
+	watcher := s.controller.WatchModelsAsUser("charlie")
+	defer workertest.CleanKill(c, watcher)
+
+	changes := watcher.Changes()
+	// discard the initial event
+	_ = s.next(c, changes)
+
+	s.ProcessChange(c, cache.ModelChange{
+		ModelUUID: "model-2-uuid",
+		Name:      "model-2",
+		Life:      life.Alive,
+		Owner:     "bob",
+		UserPermissions: map[string]permission.Access{
+			"albert":  permission.AdminAccess,
+			"bob":     permission.AdminAccess,
+			"mary":    permission.ReadAccess,
+			"charlie": permission.ReadAccess,
+		},
+	}, s.events)
+
+	update := s.next(c, changes)
+	c.Assert(update, jc.DeepEquals, []cache.ModelSummary{
+		{
+			UUID:      "model-2-uuid",
+			Namespace: "bob",
+			Name:      "model-2",
+			Admins:    []string{"albert", "bob"},
+			Status:    cache.StatusGreen,
+		},
+	})
+}
+
+func (s *modelSummaryWatcherSuite) TestRemovingPermissionRemovesModel(c *gc.C) {
+	watcher := s.controller.WatchModelsAsUser("bob")
+	defer workertest.CleanKill(c, watcher)
+
+	changes := watcher.Changes()
+	// discard the initial event
+	_ = s.next(c, changes)
+
+	s.ProcessChange(c, cache.ModelChange{
+		ModelUUID: "model-2-uuid",
+		Name:      "model-2",
+		Life:      life.Alive,
+		Owner:     "bob",
+		UserPermissions: map[string]permission.Access{
+			"mary":    permission.ReadAccess,
+			"charlie": permission.ReadAccess,
+		},
+	}, s.events)
+
+	update := s.next(c, changes)
+	c.Assert(update, jc.DeepEquals, []cache.ModelSummary{
+		{
+			UUID:    "model-2-uuid",
+			Removed: true,
+		},
+	})
+}
+
+func (s *modelSummaryWatcherSuite) TestAddModelShowsModel(c *gc.C) {
+	watcher := s.controller.WatchModelsAsUser("bob")
+	defer workertest.CleanKill(c, watcher)
+
+	changes := watcher.Changes()
+	// discard the initial event
+	_ = s.next(c, changes)
+
+	s.ProcessChange(c, cache.ModelChange{
+		ModelUUID: "model-3-uuid",
+		Name:      "model-3",
+		Life:      life.Alive,
+		Owner:     "mary",
+		UserPermissions: map[string]permission.Access{
+			"bob":  permission.ReadAccess,
+			"mary": permission.AdminAccess,
+		},
+	}, s.events)
+
+	update := s.next(c, changes)
+	c.Assert(update, jc.DeepEquals, []cache.ModelSummary{
+		{
+			UUID:      "model-3-uuid",
+			Namespace: "mary",
+			Name:      "model-3",
+			Admins:    []string{"mary"},
```

Comment:
This makes me wonder if the summary should also include non-admin users that have access to the model. (write users, and read only)

Reply:
This hasn't been asked for... but we could add it later.